### ISSUE TITLE
Managing floating window focus: Dealing with overlapping windows in GUI programs

### ISSUE THREAD

**Issue by 0xfea5** [riverwm/river#982]
I've been using River for a while now and I've noticed a small issue (or rather, a personal annoyance) when dealing with floating windows. When using GUI programs such as GIMP, it's common for both large and small floating windows to appear. With multiple floating windows, it's very easy to accidentally hover over a large one, causing it to gain focus and obscure the smaller ones beneath it. To replicate this behavior, `focus-follows-cursor` must be set to `normal` or `always`.

A simple example to illustrate this is when trying to save a file:

1) Attempt to save -> The file browser window pops up (a large floating window)
2) Select the file location and choose to overwrite an existing file -> A confirmation window pops up (a small floating window)
3) Accidentally hover over the browser window -> It gains focus and moves on top of the confirmation window.

My question is: Is there a way to address this issue? Perhaps a Wayland protocol that designates certain windows as more important, ensuring they remain on top?

Out of curiosity, I tested this on Sway, and it appears that while focus follows the cursor, windows don't actually move to the top until clicked, which, in my opinion, is a much more intuitive behavior.

**Comment by MaxVerevkin** [comment]
Relevant discussion in #775. No need for a new protocol, xdg already provides a mechanism to build parent-child relations between toplevels.

**Comment by 0xfea5** [comment]
> Relevant discussion in #775. No need for a new protocol, xdg already provides a mechanism to build parent-child relations between toplevels.

I didn't know about this mechanism so thanks for pointing it out. From what I understand though, it's in the hands of the application to designate which windows are popups and therefore they should be displayed on top of their parents. So are all these apps like GIMP (or perhaps GTK itself) not implement this properly?

**Comment by MaxVerevkin** [comment]
No, river just does not keep children views on top of parents, at least for now. Although I'm not sure if GIMP specifically sets the parent property correctly.

**Comment by maksut** [comment]
Seeing this issue with Ardour and VST plugins.

Ardour likes its floating windows (like GIMP). Focusing the floating window dialog is a hit & miss with default `focus-follows-cursor=disabled`. Because clicking on top of the floating window sometimes register in the background window (not sure if this is another defect).

So I've tried `focus-follows-cursor=normal` but then hit this issue. Tiny floating windows goes behind their "parent" windows when I accidentally move the mouse out. Happens a lot because they are tiny :)

Just another data point in case it is helpful.

Thanks for your work.

**Comment by ifreund** [comment]
> Ardour likes its floating windows (like GIMP). Focusing the floating window dialog is a hit & miss with default focus-follows-cursor=disabled. Because clicking on top of the floating window sometimes register in the background window (not sure if this is another defect).

Clicks registering on the wrong window is a bug, this sounds like https://github.com/riverwm/river/issues/833. Providing further details on that issue to help me reproduce it so I can fix it would be helpful.

Anyhow, I think the proper fix for this UX issue of the `focus-follows-cursor` feature is to decouple focus order from rendering order. I myself have never used `focus-follows-cursor` and have yet to find the motivation to work on this though.
